### PR TITLE
Add get_metric_history_bulk_interval api

### LINKED ISSUE #10455: [FR] [MLFLOW ROADMAP] Implement Metric Sampling API
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

A new feature for the MLflow UI (metric sampling for Deep Learning metric volumes during training) has been implemented and is available for use. 

In order to enable this UI functionality for MLflow backends, we would like a contributor to work on the implementation for: 
- FileStore
- SQLAlchemyStore 

and provide the functionality for handling within the MLflow tracking server. 

The RPC message definition for this API is as follows:

```python
# Returns metrics for the provided list of runs for the given step interval.
# If the number of steps exceed the max_results to return, then the resulting
# data is sampled. For example, if max_results is 5, start_step = 0, and 
# end_step = 8, then steps [0, 2, 4, 6, 8] are returned. The start_step and 
# end_step (if available) should be included in the response.
message GetMetricHistoryBulkInterval {

  # List of runs
  repeated string run_id = 1;

  # Metric key
  optional string metric_key = 2 [(validate_required) = true];

  # Number of results to return, e.g. 25k by default
  optional int32 max_results = 3;

  # Step to start returning metrics from, e.g. 0 by default
  optional int32 start_step = 4;

  # Step to stop returning metrics from, e.g. None by default (max)
  optional int32 end_step = 5;

  # Timestamp to start querying metrics for
  # - should not be set if start_step or end_step is set
  optional int64 start_timestamp = 6;

  # Timestamp to stop querying metrics for
  # - should not be set if start_step or end_step is set
  # - should be set if start_timestamp is set
  optional int64 end_timestamp = 7;

  message Response {

    # Returned metrics data
    repeated MetricWithRunId metrics = 1;

    # If the returned data was sampled
    optional bool sampled = 2;
  }
}
```

If you are interested in implementing this feature, an excellent reference point is this PR: #8739 , which shows the scope of the implementation and tests that would be required for this to be merged. 

If you are interested, please respond to this issue and we'll be sure to assign it to you. 

Note that we are interested in getting this introduced to MLflow in the 2.10.0 release (tentatively scheduled for the 2nd or 3rd week of January, 2024). If you are unable to make this delivery window due to other responsibilities, we would appreciate it if you didn't volunteer. However, if you feel like you can get this done, please let us know as we would love for this to be an external contribution! 

### Motivation

> #### What is the use case for this feature?

To support large volume metric sampling to allow for DL training workloads data to be displayed in MLflow UI charts, as well as establishing the required pre-existing work to allow for a client metric sampling API to be created.

> #### Why is this use case valuable to support for MLflow users in general?
Analyzing and evaluating extremely high volume metrics data is critical for DL training workloads (particularly for larger and more complex models that train for very long epoch iterations). 

> #### Why is this use case valuable to support for your project(s) or organization?
(see above on DL training analysis)

> #### Why is it currently difficult to achieve this use case?
The volume of data for many DL training runs is too great to load into UI visualizations with the existing APIs and is far too voluminous to be of a useful visualization data set. Deterministic statistical sampling is an effective way of visualizing extremely large metric populations.

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by serena-ruan**
### Related Issues/PRs

Resolve #10455

### What changes are proposed in this pull request?
Support GetMetricHistoryBulkInterval API
Corresponding change: https://src.dev.databricks.com/databricks/universe/-/blob/mlflow/src/main/scala/com/databricks/mlflow/MlflowBackend.scala?L1844

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1156]: can we pull this out from this function?

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1150]:
```suggestion
        steps = []
        intervals = [=1]
```

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1119]: This is an interesting requirement. Can we just return an empty array?

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1126]: do we need this casting?

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:993]:
```suggestion
            return max_step or 0
```

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:989]: What happens if `func.max` matches multiple rows?

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1197]: do we need to sort `run_ids`? can we preserve the original order?

File: mlflow/protos/service.proto
Comment (BenWilson2) [inline mlflow/protos/service.proto:339]:
```suggestion
        since { major: 2, minor: 11 },
```

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py:1119]: Databricks side enforce run_ids to be non empty

File: mlflow/server/handlers.py
Comment (BenWilson2) [inline mlflow/server/handlers.py:1177]:
```suggestion
                "If either start step or end step are specified, both must be specified."
```

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py:1126]: Yes, otherwise it's string

**Review by BenWilson2 — APPROVED** [review]
LGTM once remaining comments are addressed. We might want to think about pulling out more of those private scoped functions into a utility module to improve the maintainability of the implementation. It's beautiful code, but it's a lot to follow along with :)

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (serena-ruan) [reply mlflow/store/tracking/sqlalchemy_store.py:989]: Seems func.max can handle multiple rows with the same max step

File: mlflow/server/handlers.py
Comment (harupy) [reply mlflow/server/handlers.py:1119]: Got it :) If Databricks enforces, OSS does the same.

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py:1127]: That's a good point, maybe we should just set to MAX_RESULTS_PER_RUN, and similar for https://github.com/mlflow/mlflow/pull/11182#discussion_r1495939584 we suppress the errors.

File: mlflow/server/handlers.py
Comment (harupy) [reply mlflow/server/handlers.py:1126]: Can we use `_get_request_message` to validate and cast parameters using the proto definition?

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [reply mlflow/store/tracking/sqlalchemy_store.py:989]: Got it!

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py]: Ah ok then! I'll follow Databricks :)

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py:1141]: Should we follow Databricks behavior? It raises exception when max_results is not within the range 🤔

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1162]:
```suggestion
```

to avoid exposing something private

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1118]: /ai give me some input and output example pairs for this function.

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1176]:
```suggestion
                    "end_step must be greater than start step. "
```

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1177]:
```suggestion
                    f"Found start_step={start_step} and end_step={end_step}."
```

File: mlflow/server/handlers.py
Comment (harupy) [reply mlflow/server/handlers.py:1118]: /ai thanks, I want python code verify these examples

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py:1140]: Seems _get_request_message doesn't really check schema for GET requests with query_strings.. https://github.com/mlflow/mlflow/blob/a81ab9d461c7242a92209c4a1e5e69b15e2c6671/mlflow/server/handlers.py#L412-L415 Do we want to fix it in this PR? BTW this is not a public API, I feel like we don't even need to add proto 🤔

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py]: Those samples are not correct 🤣

Comment (harupy) [reply mlflow/server/handlers.py]: @serena-ruan Can you add a couple examples? Examples are worth 1M words.

Comment (serena-ruan) [reply mlflow/server/handlers.py]: Yes, added!

Comment (harupy) [reply mlflow/server/handlers.py]:
```
def _get_sampled_steps_from_steps(start_step, end_step, max_results):
    interval = (end_step - start_step + 1) / max_results
    return [int(start_step + i * interval) for i in range(max_results)]
```

Is this equivalent to

File: mlflow/server/handlers.py
Comment (harupy) [reply mlflow/server/handlers.py:1140]: We should use proto when we add a new endpont. It doesn't matter whether it's public or private.

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py]: It's not, for example start_step=0, end_step=100, max_results=5, interval is 20.2, then the result should be [0, 21, 41, 61, 81] while this function is [0, 20, 40, 60, 80]. I think we don't need to simplify the implementation of this function as keeping the logic same as scala code could make sure it doesn't break :)

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1169]: do we need to sort steps here?

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1218]: can we implement this as `get_metric_history_bulk_interval_from_steps` in the base store class and override it in the file and sqlalchemy store to simplify the logic here?

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1211]: Do we need this if condition? All stores implement `get_metric_history_bulk_interval_from_steps`, right?

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1111]: Does this mean the result includes both `start_step & end_step`?

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1183]: Can we validate `start_step` and `end_step` earlier?

File: mlflow/protos/service.proto
Comment (harupy) [inline mlflow/protos/service.proto:1116]: What happens if some runs are missing the specified metric?

- Run 1: has foo
- Run 2: doesn't have foo

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py:1111]: Nope, this means we need to check both start_step & end_step. But in reality, both start_step and end_step will be included in the final steps used.

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py:1211]: This is checking if `get_metric_history_bulk_interval` function exists. But I'm happy to remove it and always use our implementation, so users could only optimize `get_max_step_for_metric` and `get_metric_history_bulk_interval_from_steps` functions.

File: mlflow/protos/service.proto
Comment (serena-ruan) [reply mlflow/protos/service.proto:1116]: Then it won't be found? We're filtering by metric_key

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by harupy** [comment]
Before merge, is i

**Comment by hubertzub-db** [comment]
@harupy I did some testing and it looks ok, LMK if there is a certain dataset you'd want me to test it with

Screenshot 2024-02-21 at 11 00 15

**Review by hubertzub-db — APPROVED** [review]
Code looks good at a glance (however I'm not a python buff 🙂 ).
I tested this with UI and it seems to work ok 💪

File: mlflow/server/handlers.py
Comment (harupy) [inline mlflow/server/handlers.py:1187]:
```suggestion
            return set(sampled_steps).union(max_steps_per_run)
```

Does this work?

Comment (serena-ruan) [reply mlflow/server/handlers.py:1187]: Nope, sqlalchemy store requires steps to be list

File: mlflow/server/handlers.py
Comment (chenmoneygithub) [inline mlflow/server/handlers.py:1179]: nit: cannot

File: mlflow/server/handlers.py
Comment (chenmoneygithub) [inline mlflow/server/handlers.py:1211]: if the step is empty, what happens? We hit an issue in managed MLflow due to empty steps, so we added extra handling on the situation where the `steps` is an empty list.

**Review by chenmoneygithub — APPROVED** [review]
Thanks Serena, looks pretty good! Left a question, otherwise LGTM.

File: mlflow/server/handlers.py
Comment (serena-ruan) [reply mlflow/server/handlers.py:1211]: If `steps` is an empty list (which it actually won't be, because the returned value of `_get_sampled_steps` always unions max_steps_per_run or end_run), then `get_metric_history_bulk_interval_from_steps` will return an empty list as well.

**Pull request merged**
